File: packages/react-native-reanimated/android/src/main/java/com/swmansion/worklets/JSCallInvokerResolver.java
package com.swmansion.worklets;

import androidx.annotation.OptIn;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.common.annotations.FrameworkAPI;
import com.facebook.react.turbomodule.core.CallInvokerHolderImpl;

public class JSCallInvokerResolver {

  @OptIn(markerClass = FrameworkAPI.class)
  public static CallInvokerHolderImpl getJSCallInvokerHolder(ReactApplicationContext context) {
    try {
      var method = context.getClass().getMethod("getJSCallInvokerHolder");
      return (CallInvokerHolderImpl) method.invoke(context);
    } catch (Exception ignored) {
      // In newer implementations, the method is in CatalystInstance, continue.
    }
    try {
      var catalystInstance = context.getClass().getMethod("getCatalystInstance").invoke(context);
      assert catalystInstance != null;
      var method = catalystInstance.getClass().getMethod("getJSCallInvokerHolder");
      return (CallInvokerHolderImpl) method.invoke(catalystInstance);
    } catch (Exception e) {
      throw new RuntimeException("Failed to get JSCallInvokerHolder", e);
    }
  }
}

File: packages/react-native-worklets/android/src/main/java/com/swmansion/worklets/JSCallInvokerResolver.kt
package com.swmansion.worklets

import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.common.annotations.FrameworkAPI
import com.facebook.react.turbomodule.core.CallInvokerHolderImpl

object JSCallInvokerResolver {

    @OptIn(FrameworkAPI::class)
    @JvmStatic
    fun getJSCallInvokerHolder(context: ReactApplicationContext): CallInvokerHolderImpl {
        try {
            val method = context.javaClass.getMethod("getJSCallInvokerHolder")
            return method.invoke(context) as CallInvokerHolderImpl
        } catch (_: Exception) {
            // In newer implementations, the method is in CatalystInstance, continue.
        }
        try {
            val catalystInstance =
                context.javaClass.getMethod("getCatalystInstance").invoke(context)
            checkNotNull(catalystInstance)
            val method = catalystInstance.javaClass.getMethod("getJSCallInvokerHolder")
            return method.invoke(catalystInstance) as CallInvokerHolderImpl
        } catch (e: Exception) {
            throw RuntimeException("Failed to get JSCallInvokerHolder", e)
        }
    }
}
